src/coprocessor/dag/batch_executor/table_scan_executor.rs

```diff
@@ -0,0 +1,906 @@
+            // Do nothing
+        } else {
+            // The layout of value is: [col_id_1, value_1, col_id_2, value_2, ...]
+            // where each element is datum encoded.
+            // The column id datum must be in var i64 type.
+            let mut remaining = value;
+            while !remaining.is_empty() && decoded_columns < columns_len {
+                if remaining[0] != datum::VAR_INT_FLAG {
+                    return Err(box_err!("Unable to decode row: column id must be VAR_INT"));
+                }
+                remaining = &remaining[1..];
+                let column_id = box_try!(number::decode_var_i64(&mut remaining));
+                let (val, new_remaining) = datum::split_datum(remaining, false)?;
+                // Note: The produced columns may be not in the same length if there is error due
+                // to corrupted data. It will be handled in `ScanExecutor`.
+                let some_index = self.column_id_index.get(&column_id);
+                if let Some(index) = some_index {
+                    let index = *index;
+                    if !self.is_column_filled[index] {
+                        columns[index].push_raw(val);
+                        decoded_columns += 1;
+                        self.is_column_filled[index] = true;
+                    } else {
+                        // This indicates that there are duplicated elements in the row, which is
+                        // unexpected. We won't abort the request or overwrite the previous element,
+                        // but will output a log anyway.
+                        warn!(
+                            "Ignored duplicated row datum in table scan";
+                            "key" => log_wrappers::Key(&key),
+                            "value" => log_wrappers::Key(&value),
+                            "dup_column_id" => column_id,
+                        );
+                    }
+                }
+                remaining = new_remaining;
+            }
+        }
+
+        // Some fields may be missing in the row, we push corresponding default value to make all
+        // columns in same length.
+        for i in 0..columns_len {
+            if !self.is_column_filled[i] {
+                // Missing fields must not be a primary key, so it must be
+                // `LazyBatchColumn::raw`.
+
+                let default_value = if !self.columns_default_value[i].is_empty() {
+                    self.columns_default_value[i].as_slice()
+                } else if !self.schema[i]
+                    .flag()
+                    .contains(cop_datatype::FieldTypeFlag::NOT_NULL)
+                {
+                    datum::DATUM_DATA_NULL
+                } else {
+                    // For a normal request this will never happen, so let's return an error.
+                    return Err(box_err!(
+                        "Invalid request: default value must be provided for NOT NULL column (offset = {})",
```

Comment:
Wrong error message.

Reply:
It's a correct error message (the request is invalid)

Reply:
ok